services/shared/network-configuration/network-configuration.ts

```diff
@@ -0,0 +1,259 @@
+    }
+
+    /**
+     * @returns Array<Machine>
+     */
+    getAdjacentMachines(): Machine[] {
+        return Array.from(this.adjacentMachines.values())
+    }
+
+    /**
+     * Check if a link exists between this machine and
+     * another machines
+     * 
+     * @param otherMachine
+     * @returns Boolean
+     */
+    hasLink(otherMachine: Machine): Boolean {
+        const otherMachineId = otherMachine.machineId
+        return this.hasLinkById(otherMachineId)
+    }
+
+    /**
+     * Check if a link exists between this machine and
+     * another machines given
+     * 
+     * @param otherMachineId
+     * @returns Boolean
+     */
+    hasLinkById(otherMachineId: string): Boolean {
+        return this.adjacentMachines.has(otherMachineId)
+    }
+}
+
+
+/**
+ * Overarching definition of a computer network: an organized
+ * collection of machines in a graph.
+ */
+export class Network {
+    readonly version: number = 1 // simple sequential API versioning
+    readonly owner: string // TODO: replace this with a User type?
+    /**
+     * Unique identifier generated for the network
+     */
+    readonly networkId: string = uuidv4()
+    networkName: string
+    readonly createdAt: Date
+    updatedAt: Date
+    /**
+     * Set of all machines in the network
+     */
+    private machines: Map<string, Machine>
+
+    constructor(owner: string, networkName: string) {
+        this.owner = owner
+        this.networkName = networkName
+        this.machines = new Map<string, Machine>()
+
+        const now = new Date()
+        this.createdAt = now
+        this.updatedAt = now
+    }
+
+    /**
+     * Retrieve all machines in the network
+     */
+    getAllMachines(): Machine[] {
+        return Array.from(this.machines.values())
+    }
+
+    /**
+     * Add a new machine to the network.
+     * Does nothing if the machine already exists in the network.
+     * 
+     * @param machine
+     */
+    addMachine(machine: Machine): void {
+        const machineId = machine.machineId
+        this.machines.set(machineId, machine)
+        this.updatedAt = new Date()
+    }
+
+    /**
+     * Remove a machine from a network. Also removes the machine
+     * as an adjacent machine from the necessary machines.
+     * Does nothing if the machine doesn't exist in the network
+     * 
+     * @param machine 
+     */
+    removeMachine(machine: Machine): void {
+        const adjacentMachines = machine.getAdjacentMachines()
+
+        const machineId = machine.machineId
+        this.machines.delete(machineId)
+
+        // Remove links between machine and other machines
+        for (const adjacent of adjacentMachines) {
+            // If the adjacent machine exists in the network,
+            // remove `machine` as being adjacent to it
+            const adjacentMachineId = adjacent.machineId
+            if (this.machines.has(adjacentMachineId)) {
+                adjacent.removeAdjacentMachine(machine)
```

Comment:
Is this function safe? If it throws any errors it will break the cascade delete.

Reply:
If `machine` isn't in `adjacent`'s adjacency list? `removeAdjacentMachine(...)` will never fail even if the `machine` doesn't exist. It uses Map.delete which can't fail. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/delete